build_projects/dotnet-cli-build/TarGzFileCreateFromDirectory.cs

```diff
@@ -0,0 +1,143 @@
+
+        /// <summary>
+        /// The path of the archive to be created.
+        /// </summary>
+        [Required]
+        public string DestinationArchive { get; set; }
+
+        /// <summary>
+        /// Indicates if the destination archive should be overwritten if it already exists.
+        /// </summary>
+        public bool OverwriteDestination { get; set; }
+
+        /// <summary>
+        /// If zipping an entire folder without exclusion patterns, whether to include the folder in the archive.
+        /// </summary>
+        public bool IncludeBaseDirectory { get; set; }
+
+        /// <summary>
+        /// An item group of regular expressions for content to exclude from the archive.
+        /// </summary>
+        public ITaskItem[] ExcludePatterns { get; set; }
+
+        protected override bool ValidateParameters()
+        {
+            base.ValidateParameters();
+
+            var retVal = true;
+
+            if (File.Exists(DestinationArchive))
+            {
+                if (OverwriteDestination == true)
+                {
+                    Log.LogMessage(MessageImportance.Low, $"{DestinationArchive} will be overwritten");
+                }
+                else
+                {
+                    Log.LogError($"'{DestinationArchive}' already exists. Did you forget to set '{nameof(OverwriteDestination)}' to true?");
+
+                    retVal = false;
+                }
+            }
+
+            SourceDirectory = Path.GetFullPath(SourceDirectory);
+
+            SourceDirectory = SourceDirectory.EndsWith(Path.DirectorySeparatorChar.ToString())
+                ? SourceDirectory
+                : SourceDirectory + Path.DirectorySeparatorChar;
+
+            if (!Directory.Exists(SourceDirectory))
+            {
+                Log.LogError($"SourceDirectory '{SourceDirectory} does not exist.");
+
+                retVal = false;
+            }
+
+            return retVal;
+        }
+
+        public override bool Execute()
+        {
+            return base.Execute();
+        }
+
+        protected override string ToolName
+        {
+            get { return "tar"; }
+        }
+
+        protected override MessageImportance StandardOutputLoggingImportance
+        {
+            get { return MessageImportance.High; } // or else the output doesn't get logged by default
+        }
+
+        protected override string GenerateFullPathToTool()
+        {
+            return "tar";
+        }
+
+        protected override string GenerateCommandLineCommands()
+        {
+            return $"{GetDestinationArchive()} {GetSourceSpecification()}";
+        }
+
+        private string GetSourceSpecification()
+        {
+            if (IncludeBaseDirectory)
+            {
+                var parentDirectory = Directory.GetParent(SourceDirectory).Parent.FullName;
+
+                var sourceDirectoryName = Path.GetFileName(Path.GetDirectoryName(SourceDirectory));
```

Comment:
I don't think this is correct.  If "SourceDirectory" contains a trailing slash then Path.GetDirectoryName returns that same directory.

[https://msdn.microsoft.com/en-us/library/system.io.path.getdirectoryname%28v=vs.110%29.aspx](https://msdn.microsoft.com/en-us/library/system.io.path.getdirectoryname%28v=vs.110%29.aspx)
`GetDirectoryName('C:\MyDir\') returns 'C:\MyDir'`